### PR TITLE
chore: Update AVL Tree

### PR DESCRIPTION
**Pull request by ritk20**
#### Description of Change

Updated AVL Tree data structure using modern C++ STL library features like `std::unique_ptr` and implemented Object Oriented structure. Also added additional tests and updated existing tests to use `assert`.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
Completed the following TODO
```
/**
 * \todo update code to use C++ STL library features and OO structure
 * \warning This program is a poor implementation and does not utilize any of
 * the C++ STL features.
 * /
```

### CODE REVIEW

File: data_structures/avltree.cpp [hunk 1]
```diff
@@ -1,198 +1,426 @@
 /**
  * \file
- * \brief A simple tree implementation using nodes
- *
- * \todo update code to use C++ STL library features and OO structure
- * \warning This program is a poor implementation and does not utilize any of
- * the C++ STL features.
+ * \brief An implementation of AVL tree.
```
Comment (realstealthninja) [inline data_structures/avltree.cpp:3]: please also update the algorithm to use @'s instead of \ for conistency

File: data_structures/avltree.cpp [hunk 2]
```diff
@@ -1,198 +1,426 @@
...
+#include <sstream>    /// for std::ostringstream
+#include <vector>     /// for std::vector
 
-using node = struct node {
-    int data;
-    int height;
-    struct node *left;
-    struct node *right;
-};
+class AVLTree {
```
Comment (realstealthninja) [inline data_structures/avltree.cpp:13]: classes should be documented check [contribution guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)

File: data_structures/avltree.cpp [hunk 3]
```diff
@@ -1,198 +1,426 @@
...
+ public:
+    struct Node {
```
Comment (realstealthninja) [inline data_structures/avltree.cpp:15]: this struct could be external to the class, it should also be documented

File: data_structures/avltree.cpp [hunk 4]
```diff
@@ -1,198 +1,426 @@
...
+        return current;
     }
-    // Balancing Tree after deletion
-    return root;
+};
+
+/**
+ * @brief Function for testing insert().
+ *
+ * @returns `void`
```
Comment (realstealthninja) [inline data_structures/avltree.cpp:219]:
```suggestion
 * @returns void
```

File: data_structures/avltree.cpp [hunk 5]
```diff
@@ -1,198 +1,426 @@
...
-    std::cout << "LevelOrder: ";
-    levelOrder(root);
-    root = deleteNode(root, 1);  // Deleting key with value 1
-    std::cout << "\nLevelOrder: ";
-    levelOrder(root);
-    root = deleteNode(root, 4);  // Deletin key with value 4
-    std::cout << "\nLevelOrder: ";
-    levelOrder(root);
-    deleteAllNodes(root);
+    test_insert();
```
Comment (realstealthninja) [inline data_structures/avltree.cpp:419]: these tests should be called by a single function named `tests()`

File: data_structures/avltree.cpp [hunk 6]
```diff
@@ -1,198 +1,426 @@
...
+    test_deleteNode();
+    test_levelOrder();
+    test_balancing();
+    test_edge_cases();
+    std::cout << "All tests passed!" << std::endl;
     return 0;
-}
+}
```
Comment (realstealthninja) [inline data_structures/avltree.cpp:426]:
```suggestion
}

```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by realstealthninja** [comment]
this pr should be put on hold until #2746

File: data_structures/avltree.cpp (hunk 6 above)
Comment (ritk20) [reply data_structures/avltree.cpp]: did not understand this issue... Is it resolved?

**Comment by ritk20** [comment]
Is it necessary to add class names and variables in snake_case as well?

**Comment by realstealthninja** [comment]
> Is it necessary to add class names and variables in snake_case as well?

No, the contributing.md file prefers pascal or camel case last i checked please do check and confirm

**Comment by ritk20** [comment]
> > Is it necessary to add class names and variables in snake_case as well?
> 
> No, the contributing.md file prefers pascal or camel case last i checked please do check and confirm

For classes, variables, and functions, there is no explicit mention of a case preference, although the description [below](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) led me to believe that snake_case is preferred.
```
/**
 * @brief Class documentation
 */
class class_name {
 private:
    int variable;   ///< short info of this variable
    char *message;  ///< short info

 public:
    // other members should be also documented as below
}

```

Also, this makes me believe that snake_case is preferred for functions
```
template <typename T>
bool is_number_on_array(const std::vector<T> &arr, const int &number) {
    for (int i = 0; i < sizeof(arr) / sizeof(int); i++) {
        if (arr[i] == number) {
            return true;
        }
        else {
            // Number not in the current index, keep searching.
        }
    }

    return false;
}
```

**Comment by realstealthninja** [comment]
feel free to use snake case instead

**Pull request closed without merging**